teos/src/responder.rs

```diff
@@ -548,69 +407,51 @@ impl chain::Listen for Responder {
-            let (_, rejected_trackers) = self.rebroadcast(self.get_txs_to_rebroadcast(height));
-            // Delete trackers rejected during rebroadcast
-            let trackers_to_delete_gk = rejected_trackers
-                .iter()
-                .map(|uuid| (*uuid, self.trackers.lock().unwrap()[uuid].user_id))
-                .collect();
-            self.delete_trackers(
-                &rejected_trackers,
-                &self
-                    .gatekeeper
-                    .delete_appointments_from_memory(&trackers_to_delete_gk),
-                DeletionReason::Rejected,
-            );
+        // We might be connecting a new block after a disconnection (reorg).
+        // We will need to update those trackers that have been reorged.
+        let did_reorg = !self.reorged_trackers.lock().unwrap().is_empty();
+        if did_reorg {
+            // Handle reorged transactions. This clears `self.reorged_trackers`.
+            self.handle_reorged_txs(height);
+        }
 
-            // Remove all receipts created in this block
-            self.carrier.lock().unwrap().clear_receipts();
+        // Delete trackers completed at this height
+        let completed_trackers = self.check_confirmations(txs.keys().cloned().collect(), height);
+        if !completed_trackers.is_empty() {
+            self.gatekeeper
+                .delete_appointments(completed_trackers, true);
+        }
 
-            if self.trackers.lock().unwrap().is_empty() {
-                log::info!("No more pending trackers");
-            }
+        // Rebroadcast those transactions that need to
+        let rejected_trackers = self.rebroadcast_stale_txs(height);
+        if !rejected_trackers.is_empty() {
+            self.gatekeeper
+                .delete_appointments(rejected_trackers, false);
         }
+
+        // Remove all receipts created in this block
+        self.carrier.lock().unwrap().clear_receipts();
     }
 
     /// Handles reorgs in the [Responder].
     fn block_disconnected(&self, header: &BlockHeader, height: u32) {
         log::warn!("Block disconnected: {}", header.block_hash());
+        // Update the carrier and our tx_index.
         self.carrier.lock().unwrap().update_height(height);
         self.tx_index
             .lock()
             .unwrap()
             .remove_disconnected_block(&header.block_hash());
-
-        for tracker in self.trackers.lock().unwrap().values_mut() {
-            // The transaction has been unconfirmed. Flag it as reorged out so we can rebroadcast it.
-            if tracker.status == ConfirmationStatus::ConfirmedIn(height) {
-                tracker.status = ConfirmationStatus::ReorgedOut;
-            }
-        }
+        // And store the reorged transactions to be retried later.
+        // TODO: Not only confirmed trackers need to be marked as reorged, but trackers that hasn't confirmed but their
+        // dispute did confirm in the reorged block.
```

Comment:
Add this note: We can either store the dispute tx confirmation, or pull the dispute txid and check it against our txindex.